File: frontend/src/services/incident.service.ts
import apiClient from "./api";
import { Incident } from "@/types";

export const incidentService = {
  getIncidents: async (): Promise<Incident[]> => {
    const response = await apiClient.get<Incident[]>("/api/v1/incidents");
    return response.data;
  },

  getIncidentById: async (id: string): Promise<Incident> => {
    const response = await apiClient.get<Incident>(`/api/v1/incidents/${id}`);
    return response.data;
  },

  getIncidentByNumber: async (incidentNumber: string): Promise<Incident> => {
    const response = await apiClient.get<Incident>(
      `/api/v1/incidents/number/${incidentNumber}`,
    );
    return response.data;
  },

  createIncident: async (data: {
    incidentNumber: string;
    title: string;
    description: string;
    priority: "CRITICAL" | "HIGH" | "MEDIUM" | "LOW";
    reporter: string;
  }): Promise<Incident> => {
    const response = await apiClient.post<Incident>("/api/v1/incidents", data);
    return response.data;
  },

  updateIncidentStatus: async (
    id: string,
    status: "NEW" | "ASSIGNED" | "IN_PROGRESS" | "PENDING" | "RESOLVED" | "CLOSED" | "REOPENED",
  ): Promise<Incident> => {
    const response = await apiClient.put<Incident>(
      `/api/v1/incidents/${id}/status?status=${status}`,
    );
    return response.data;
  },

  assignIncident: async (id: string, assignee: string): Promise<Incident> => {
    const response = await apiClient.put<Incident>(
      `/api/v1/incidents/${id}/assign?assignee=${assignee}`,
    );
    return response.data;
  },

  deleteIncident: async (id: string): Promise<void> => {
    await apiClient.delete(`/api/v1/incidents/${id}`);
  },
};

File: itsm/src/main/java/com/snowrepo/itsm/application/IncidentService.java
package com.snowrepo.itsm.application;

import com.snowrepo.itsm.domain.Incident;
import com.snowrepo.itsm.domain.Incident.IncidentStatus;
import com.snowrepo.itsm.domain.Incident.Priority;
import com.snowrepo.itsm.domain.IncidentRepository;
import java.util.List;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional
@RequiredArgsConstructor
@Slf4j
@SuppressWarnings("null")
public class IncidentService {

  private final IncidentRepository incidentRepository;

  public Incident createIncident(
      String incidentNumber,
      String title,
      String description,
      Priority priority,
      String reporter,
      String createdBy
  ) {
    log.info("Creating incident: {}", incidentNumber);

    Incident incident = Incident.create(
        incidentNumber,
        title,
        description,
        priority,
        reporter,
        createdBy
    );

    return incidentRepository.save(incident);
  }

  public Incident getIncident(UUID id) {
    log.info("Fetching incident: {}", id);
    return incidentRepository.findById(id)
        .orElseThrow(() -> new RuntimeException("Incident not found: " + id));
  }

  public Incident getIncidentByNumber(String incidentNumber) {
    log.info("Fetching incident by number: {}", incidentNumber);
    return incidentRepository.findByIncidentNumber(incidentNumber)
        .orElseThrow(() -> new RuntimeException("Incident not found: " + incidentNumber));
  }

  public List<Incident> getAllIncidents() {
    log.info("Fetching all incidents");
    return incidentRepository.findAll();
  }

  public Incident updateIncidentStatus(UUID id, IncidentStatus newStatus, String updatedBy) {
    log.info("Updating incident status: {} -> {}", id, newStatus);

    Incident incident = getIncident(id);
    incident.updateStatus(newStatus, updatedBy);

    return incidentRepository.save(incident);
  }

  public Incident assignIncident(UUID id, String assignee, String updatedBy) {
    log.info("Assigning incident {} to {}", id, assignee);

    Incident incident = getIncident(id);
    incident.assignTo(assignee, updatedBy);

    return incidentRepository.save(incident);
  }

  public void deleteIncident(UUID id) {
    log.info("Deleting incident: {}", id);
    incidentRepository.deleteById(id);
  }
}
